### DISCUSSION TITLE
Do you find it easy to create themes and custom UI?

### DISCUSSION THREAD

**Discussion by ldimitrijevic** [OrchardCMS/OrchardCore#12970] [General]
Hello,
first of all I would like to thank all contributors to the project which I find very interesting. We have setup an Orchard Core CMS for a client and it has been quite a struggle, but we succeeded.
As I wanted to deep dive and understand fully this CMS I started a personal project: I've found myself a design: https://demo.templatemonster.com/demo/157393.html and wanted to integrate it with Orchard Core. Was it a mistake? I don't know. I've followed all Lombiq tutorials on youtube but still there are some pieces missing.
My first objective was to implement a section header:
Title, Subtitle, Description, an image for the section and two other parameters: a boolean value and another one.
A section should be contained in a Page and therefore I added a BagPart to this "Page" content item.
Why a bag part? Because flow is to me useless. You can configure "column" like style but in the front part, there is only a class on the html tag telling it is 50% for instance. How do you make this responsive? With a lot of work (too much) and with no satisfying result. The other option was then to have a BagPart and to override almost everything. I did this with liquid template. I achieved it with a lot of effort. Nothing is really natural. The naming of the templates is sometimes strange, the detail and summary views can be very confusing etc. Nonetheless I did it, but when it came to another bag part to handle the content of the section, all was to be done again as I had already overridden the bag part template.
I could have stopped there but i decided to write a module and make my own part. Drivers are not that difficult to write after all. But here again the template mechanism and naming appeared to be very strange. To build my section Item I added several parts:

```
_contentDefinitionManager.AlterPartDefinition(nameof(SectionTitlePart), part => part
            .Attachable()
            .Reusable()
            .WithField(nameof(SectionTitlePart.Title), field => field
                .OfType(nameof(TextField))
                .WithDescription("Titre de la section")
                .WithDisplayName("Titre"))
            .WithField(nameof(SectionTitlePart.Subtitle), field => field
                .OfType(nameof(HtmlField))
                .WithDescription("Sous-titre de la section")
                .WithDisplayName("Sous-titre")
                .WithSettings<HtmlFieldSettings>(new HtmlFieldSettings
                {
                    Hint = "Pour mettre un élément en évidence, encadrez-le d'une balise <span>",
                    SanitizeHtml = true
                }))
        );

        _contentDefinitionManager.AlterPartDefinition(nameof(SectionDescriptionPart), part => part
            .Attachable()
            .Reusable()
            .WithField(nameof(SectionDescriptionPart.Description), field => field
                .OfType(nameof(TextField))
                .WithDescription("Description de la section")
                .WithDisplayName("Description")
                .WithEditor("TextArea")
            ));

        _contentDefinitionManager.AlterPartDefinition(nameof(SectionImagePart), part => part
            .Attachable()
            .Reusable()
            .WithField(nameof(SectionImagePart.Image), field => field
                .OfType(nameof(MediaField))
                .WithDescription("Image de la section")
                .WithDisplayName("Image")
            ));

        _contentDefinitionManager.AlterPartDefinition(nameof(SectionDisplayOptionsPart), part => part
            .Attachable()
            .Reusable()
            .WithField(nameof(SectionDisplayOptionsPart.TwoColunmsHeader), field => field
                .OfType(nameof(BooleanField))
                .WithDescription("En-tête avec deux colonnes")
                .WithDisplayName("En-tête avec deux colonnes")
                .WithEditor("Switch")
                .WithSettings<BooleanFieldSettings>(new BooleanFieldSettings
                {
                    DefaultValue = false,
                    Label = "Afficher l'en-tête sur deux colonnes",
                    Hint = "Lorsque cette option est activée, la partie titre et la partie description sont affichées côte à côte"
                })
            ));

        _contentDefinitionManager.AlterTypeDefinition("Section", type => type
            .Creatable()
            .Listable()
            .WithPart(nameof(SectionTitlePart))
            .WithPart(nameof(SectionDescriptionPart))
            .WithPart(nameof(SectionImagePart))
            .WithPart(nameof(SectionDisplayOptionsPart))
           );

```
Nothing very difficult but here again, the template system is incomprehensible. As i decided to have a better admin UX, as i was already building everything I decided to design things. So to have the design of the section ContentType I had to add a Content-Section.Edit.cshtml file in Items folder. Why there? I don't know, i just followed Lombiq examples. Anyhow moving this file out of this folder doesn't make it work. The html is very simple:

```
<div class="row">
    <div class="col-6">
        @await DisplayAsync(Model.Parts["SectionTitlePart"])
    </div>
    <div class="col-6">
        @await DisplayAsync(Model.Parts["SectionDescriptionPart"])
    </div>
</div>
<div class="row">
    <div class="col-6">
        @await DisplayAsync(Model.Parts["SectionImagePart"])
    </div>
    <div class="col-6">
        <div class="section-options d-flex flex-column">
            <div class="mb-3">Options d'affichage</div>
            @await DisplayAsync(Model.Parts["SectionDisplayOptionsPart"])
           
        </div>
        
    </div>
</div>

<div class="edit-item-actions">
    @await DisplayAsync(Model.Actions)

    @{
        var returnUrl = Context.Request.Query["returnUrl"];
    }

    @if (!String.IsNullOrWhiteSpace(returnUrl) && Url.IsLocalUrl(returnUrl))
    {
        <a class="btn btn-secondary cancel" role="button" href="@returnUrl">@T["Cancel"]</a>
    }
</div>
```

I also copied the Publish and other Buttons at the end which made me duplicate the code... Strange. I was happy to find that I could render parts by their name even if I don't like the hard coded string (maybe I could use a nameof but i didn't try).

My admin part was fine ! Two columns, things could be saved and edited, it was now the time to build the front part.

Here again, strange things with template namings.... I had to create (in the items folder) a Content-Section.Display.cshtml and thought that the same mechanism used to build the edit part could be used... Unfortunately, the model is not the same and there is no "Parts" collection...

Here is my story until now. If someone can help I would be glad and thankful.

The lesson of all this "adventure" is that the learning curve is very important. Orchard CMS is nice if you are a developer but it is not made for end users. And it is a big deal ! You build CMS to give a non developer the responsibility to build its own web site. The documentation is not sufficient if you want to go further than simply adding a few HTML Fields and media. As soon as you want something more sophisticated you have to struggle with the system.
These are a few thoughts of mine but I've read others also telling, the system should be closer to what WordPress is. That is, something easy to manipulate for a non developer.

Anyhow thanks to the developers for all the work that has been done until now and I hope I will tackle my problems without rebuilding the whole system :)

**Comment by ns8482e** [comment]
@ldimitrijevic OOTB Orchard core provides you admin editing experience for all your custom parts/fields defined. That covers most users editing scenarios without adding alternates to `Content_Edit` shape.  In most scenarios, a user defines Fileds/Parts via admin, and defines  Views for Display and OC takes care of Edit.

Here, by defining alternate `Content-Section.Edit.cshtml` you are telling orchard core system that you are not using OOTB Edit ('Content_Edit') for specific content type `Section` and you are building your own parallel edit experience for content type `Section` -  OC gives you such power and allow you to define your own edit templates using alternates to customize edit experience.

However that comes with a cost - as you are not leveraging OOTB Edit for 'Section' and you may need to copy and duplicate code as needed per your customized edit experience needs.

`Views` and `Views\Items` are view locations from which OC discovers your views

**Reply by ldimitrijevic** [reply]
Hello,
thanks for the reply. I've overridden because I think that UX is better with two columns than with a single one. Just to explain: When you create a "Page" content item, and let the end-user select the items he wishes, each field comes one below the other. That makes quite long forms. As some of the fields are not meant to be long i prefered creating a "two columns" experience. I globally understand the issues (though sometimes i wonder if i do not miss a case), but i still think it is to complicated for such "simple" use cases. To be clear, the difficulty is not on the code or on writting the template but to guess how the template should be named and what model is passed to it. The only way i've found was by experiencing which takes long...

My admin part works fine now, but the issue is on the front part. And the biggest strange thing is that front and back parts do not work the same. I mean, they do not have similar models. The paradox is that, if i do not override the views, the engine outputs the parts twice... If i override the Section view, i'm stuck on the front as I can not access the parts as i did in the back... The Model.Parts property does not exists. There is an Items property though but the only way to access items seams to be by looping through the collection and then display the part i want. This is working but it is neither elegant nor efficient.

I find the principle of drivers very interesting and very powerful but i wonder why there are no drivers for content types. There are drivers for fields, for parts and if there was a driver for content types, my "issue" should be solved easily.

> Here, by defining alternate Content-Section.Edit.cshtml you are telling orchard core system that you are not using OOTB Edit ('Content_Edit') for specific content type Section and you are building your own parallel edit experience for content type Section - OC gives you such power and allow you to define your own edit templates using alternates to customize edit experience.

I agree but... When you create a part with fields and with simple properties you manage yourself, fields are automatically added to the output. I mean i do not have to explicitely output them in my view (though i think it would give even more flexibility to the developer, allowing him to display the content as he likes). Therfore i thought, that the common buttons of the admin would also be automatically generated which is not the case.

> Views and Views\Items are view locations from which OC discovers your views

Unfortunately, i did not find any documentation on the Views/Items folder (found it in the orchard core code).

By the way, any idea on how to solve my issue on the front? You would be my Santa for 2022 :)

**Comment by ldimitrijevic** [comment]
Just to add a word, can you explain why on the admin, the Content-Section.Edit.cshtml file has to be in the Views/Items folder and why on the front, the Content-Section.Summary.cshtml has to be in the Views folder, and also why summary and why not Detail or Display or simply Content-Section. It is really hard to understand when and why it is one display or another. I think summary is in a list but what if all my displays are the same? Shall i duplicate all of them?

**Comment by ldimitrijevic** [comment]
I succeed :) though i do not like the solution that much.
I specified in the part the "view" i want and positionned it in a specific zone. That way i am able to tell the section template  what part i want in the html structure:

The driver:

```
public override IDisplayResult Display(SectionTitlePart part, BuildPartDisplayContext context)
    {
        //return Initialize<SectionTitlePartViewModel>(
        //    GetDisplayShapeType(context),
        //    viewModel => PopulateViewModel(part, viewModel)).Location("SectionTitle");

        return View("SectionTitle_Display", part).Location("SectionTitle");
    }
```

The Section template:

```
<div class="section">
    @await DisplayAsync(Model.SectionTitle)
</div>
```

This is of course a sample and not the definitive HTML structure.
And the Title part template (it has to be named SectionTitle.Display.cshtml within View/Items folder):

```
@model ShapeViewModel<PTA.Modules.Section.Models.SectionTitlePart>

<div class="section-title">
    <h5 class="title">@Model.Value.Title.Text</h5>
    <h2 class="subtitle">@Html.Raw(@Model.Value.Subtitle.Html)</h2>
</div>
```

If there is a cleaner solution let me know.
